### PR TITLE
kubelet: Prevent runtime-only pods from going into terminated phase

### LINKED ISSUE #103625: Kubelet panic in syncTerminatedPod
#### What happened:

When testing #102344 against OpenShift CI, I discovered a panic in the kubelet: https://github.com/openshift/kubernetes/pull/851#issuecomment-877469501

Further investigation revealed that it is possible for `syncTerminatedPod` to panic under certain circumstances: https://github.com/kubernetes/kubernetes/blob/a0d77e5763ac04ddc713bf2e284580d32fd4b079/pkg/kubelet/pod_workers.go#L685-L713

We hit this case when `update.Options.RunningPod != nil` (podStatus remains nil) and `update.WorkType == TerminatedPodWork` (nil will be passed along to `syncTerminatedPod`). I'm not sure why exactly that would happen but it seems possible.

#### What you expected to happen:

Don't panic!

#### How to reproduce it (as minimally and precisely as possible):

Unclear, I've been trying to trace the code/find evidence of this happening in k8s CI but haven't found a match yet. (https://github.com/kubernetes/test-infra/issues/22851 would be nice :smile:)

#### Anything else we need to know?:

/sig node
/cc @rphillips @smarterclayton 

/priority important-soon
/milestone v1.22

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by smarterclayton**
If a pod is already in terminated and the housekeeping loop sees an
out of date cache entry for a running container, the pod worker
should ignore that running pod termination request. Once the worker
completes, a subsequent housekeeping invocation will then invoke
terminating because the worker is no longer processing any pod
with that UID.

This does leave the possibility of syncTerminatedPod being blocked
if a container in the pod is started after killPod successfully
completes but before syncTerminatedPod can exit successfully,
perhaps because the terminated flow (detach volumes) is blocked on
that running container. A future change will address that issue.

#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #103625

```release-note
NONE
```

```docs
```

### CODE REVIEW

File: pkg/kubelet/pod_workers.go [hunk 1]
```diff
@@ -537,6 +537,14 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 	var wasGracePeriodShortened bool
 	switch {
 	case status.IsTerminated():
+		// A terminated pod may still be waiting for cleanup - if we recieve a runtime pod kill request
+		// due to housekeeping seeing an older cached version of the runtime pod simply ignore it until
+		// after the pod worker completes.
+		if isRuntimePod {
+			klog.InfoS("Pod is waiting for termination, ignoring runtime-only kill until after pod worker is fully terminated", "pod", klog.KObj(pod), "podUID", pod.UID)
```
Comment (rphillips) [inline pkg/kubelet/pod_workers.go:544]: Once we validate this works, this log line should be at level 4 or so.

Comment (smarterclayton) [reply pkg/kubelet/pod_workers.go:544]: This should be so rare that I'd prefer to have it at 1, possibly at 2.

**Comment by Random-Liu** [comment]
/lgtm
/approve
/hold

Please address the comment. :)

File: pkg/kubelet/pod_workers.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/kubelet/pod_workers.go]: Set it at 3 (in between).

**Comment by smarterclayton** [comment]
Updated

**Comment by ehashman** [comment]
/milestone v1.22

Fixes a kubelet panic, release-critical bug.

**Review by ehashman — COMMENTED** [review]
/lgtm

**Comment by smarterclayton** [comment]
The failure appears to be related to something this change would catch (a failure to mount and then a failure to delete) within 5m.  Why would container setup fail twice in a row for that long?

```
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:56:51 +0000 UTC - event for pvc-tester-8wrw4: {kubelet e2e-75151d6a56-a7d53-minion-group-3stc} Started: Started container write-pod
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:56:51 +0000 UTC - event for pvc-tester-8wrw4: {kubelet e2e-75151d6a56-a7d53-minion-group-3stc} Created: Created container write-pod
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:56:51 +0000 UTC - event for pvc-tester-8wrw4: {kubelet e2e-75151d6a56-a7d53-minion-group-3stc} Pulled: Container image "k8s.gcr.io/e2e-test-images/busybox:1.29-1" already present on machine
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:56:52 +0000 UTC - event for pvc-tester-8wrw4: {kubelet e2e-75151d6a56-a7d53-minion-group-3stc} FailedMount: MountVolume.SetUp failed for volume "kube-api-access-clzcg" : object "pv-5286"/"kube-root-ca.crt" not registered
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:56:54 +0000 UTC - event for pvc-tester-ntf66: {default-scheduler } Scheduled: Successfully assigned pv-5286/pvc-tester-ntf66 to e2e-75151d6a56-a7d53-minion-group-c95w
Jul 13 16:02:44.739: INFO: At 2021-07-13 15:58:57 +0000 UTC - event for pvc-tester-ntf66: {kubelet e2e-75151d6a56-a7d53-minion-group-c95w} FailedMount: Unable to attach or mount volumes: unmounted volumes=[volume1], unattached volumes=[volume1 kube-api-access-g4hgx]: timed out waiting for the condition
Jul 13 16:02:44.739: INFO: At 2021-07-13 16:02:29 +0000 UTC - event for pvc-tester-ntf66: {kubelet e2e-75151d6a56-a7d53-minion-group-c95w} Pulled: Container image "k8s.gcr.io/e2e-test-images/busybox:1.29-1" already present on machine
Jul 13 16:02:44.739: INFO: At 2021-07-13 16:02:29 +0000 UTC - event for pvc-tester-ntf66: {kubelet e2e-75151d6a56-a7d53-minion-group-c95w} Created: Created container write-pod
```

**Comment by ehashman** [comment]
/retest

flakes

**Pull request merged**